### PR TITLE
DNM: quick-fixes

### PR DESCRIPTION
**Pull request by kasperpeulen**
testable here: http://kasperpeulen.github.io/dart-pad/

### CODE REVIEW

File: lib/images/bulb.svg [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg
+        xmlns:svg="http://www.w3.org/2000/svg"
+        xmlns="http://www.w3.org/2000/svg"
+        viewBox="0 0 1024 1024"
+        width="12"
+        height="12">
+
+    <path fill="#dcdccc" class="path1" d="M1002.934 817.876l-460.552-394.76c21.448-40.298 33.618-86.282 33.618-135.116 0-159.058-128.942-288-288-288-29.094 0-57.172 4.332-83.646 12.354l166.39 166.39c24.89 24.89 24.89 65.62 0 90.51l-101.49 101.49c-24.89 24.89-65.62 24.89-90.51 0l-166.39-166.39c-8.022 26.474-12.354 54.552-12.354 83.646 0 159.058 128.942 288 288 288 48.834 0 94.818-12.17 135.116-33.62l394.76 460.552c22.908 26.724 62.016 28.226 86.904 3.338l101.492-101.492c24.888-24.888 23.386-63.994-3.338-86.902z"></path>
```
Comment (lukechurch) [inline lib/images/bulb.svg]: This is actually a spanner right? Not a bulb?

File: lib/images/bulb.svg [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+</svg>
+        <!-- TODO check if there are license problems -->
```
Comment (lukechurch) [inline lib/images/bulb.svg]: It came from here https://github.com/google/material-design-icons? Could you just point to the place in the repo where it is from? 

I think it is fine in this case, we just need to add an attribution

File: lib/services/common.dart [hunk 3]
```diff
@@ -4,9 +4,9 @@
 
 library dart_pad.common;
 
-final String serverURL = 'https://dart-services.appspot.com/';
-//final String serverURL = 'http://doc-test.dart-services.appspot.com/';
-//final String serverURL = 'http://127.0.0.1:8082/';
+//final String serverURL = 'https://dart-services.appspot.com/';
+final String serverURL = 'http://doc-test.dart-services.appspot.com/';
```
Comment (lukechurch) [inline lib/services/common.dart]: Please revert this, and we'll need to do a server urpgrade before deploy the version.

**Comment by lukechurch** [comment]
Sadly, we're not going to be able to take this until https://code.google.com/p/dart/issues/detail?id=23103 gets fixed.

@scheglov

File: lib/completion.dart [hunk 4]
```diff
@@ -39,81 +39,102 @@ class DartCompleter extends CodeCompleter {
...
+              cursorPos = text.indexOf('(') + 1;
+            }
 
-          return new Completion(text, displayString: displayString,
-              type: "type-${completion.type.toLowerCase()}${deprecatedClass}",
-              cursorOffset: cursorPos);
-        }
-      }).where((x) => x != null).toList();
+            return new Completion(text, displayString: displayString,
+            type: "type-${completion.type.toLowerCase()}${deprecatedClass}",
```
Comment (lukechurch) [inline lib/completion.dart]: This indentation looks a bit wonky to me, but I'll defer to @devoncarew who is the master of ws fixes.

File: lib/editing/editor_codemirror.css [hunk 5]
```diff
@@ -177,6 +177,17 @@ li.CodeMirror-hint-active {
...
 }
 
+.type-quick_fix {
+  font-family: "Inconsolata", sans-serif;
+  font-size: 12pt;
+}
+
+.type-quick_fix::before {
+  content: " ";
+  background: url('../images/bulb.svg');
```
Comment (lukechurch) [inline lib/editing/editor_codemirror.css]: please rename the file to wrench.svg

File: lib/images/bulb.svg (hunk 2 above)
Comment (lukechurch) [inline lib/images/bulb.svg]: Providence reported as: https://icomoon.io/app/#/select search for 'wrench'. Available under GPL or CC BY 4.0.

We would need to be using it under CC BY 4.0, and would need attribution text.

I would prefer if we could find something that was easier for us to consume, e.g. the Chromimum wrench icon. @devoncarew do you know where that lives?

File: lib/playground.dart [hunk 6]
```diff
@@ -740,4 +758,4 @@ class InlineBrackets extends markdown.InlineSyntax {
     parser.addNode(element);
     return true;
   }
-}
+}
```
Comment (lukechurch) [inline lib/playground.dart]: Line  break at the end of the file please

**Comment by kasperpeulen** [comment]
Hm.. the old wrench seems to look a little bit better to me tbh:


vs the new one:

File: lib/dartservices_client/v1.dart [hunk 7]
```diff
@@ -1,3 +1,5 @@
+// This is a generated file (see the discoveryapis_generator project).
```
Comment (devoncarew) [inline lib/dartservices_client/v1.dart]: This is already a pretty big PR; let's land the generated discovery API in another PR (once it lands in the dart-services project), and update this PR from master.

File: lib/editing/editor_codemirror.dart [hunk 8]
```diff
@@ -245,6 +286,33 @@ class _CodeMirrorDocument extends Document {
...
 
+  void applyEdit(Edit edit) {
+    doc.replaceRange(
+        edit.replacement,
+        _posToPos(posFromIndex(edit.offset)),
+        _posToPos(posFromIndex(edit.offset + edit.length))
+    );
+  }
+
+  bool get hasIssueAtOffset {
```
Comment (devoncarew) [inline lib/editing/editor_codemirror.dart]: I don't think we should be storing analysis state in the editor.

File: lib/editing/editor.dart [hunk 9]
```diff
@@ -7,6 +7,7 @@ library editor;
 import 'dart:async';
 import 'dart:html' as html;
 import 'dart:math';
+import 'package:dart_pad/dartservices_client/v1.dart';
```
Comment (devoncarew) [inline lib/editing/editor.dart]: We want to avoid having the editor library depend on the services one. The editor library should be a bit more abstract than that. I think we want an implementation of a `SourceEdit` like class to live in this library. It would be similar to the services `Edit` class. People using an `Editor` instead that had made a services call would need to convert between the `Edit` class and the `SourceEdit` one; similar to what we do for code completions.

File: lib/editing/editor.dart [hunk 10]
```diff
@@ -37,6 +38,8 @@ abstract class Editor {
    */
   void execCommand(String name);
 
+  void autoComplete({bool autoInvoked, bool quickFix});
```
Comment (devoncarew) [inline lib/editing/editor.dart]: We should provide default values for these `bool`s; that way we don't have to check `if (autoInvoked != null && autoInvoked)` and similar. And perhaps `autoComplete` ==> `showCompletions`? That way there isn't ambiguity between autoComplete and autoInvoked. Lastly, I think the intent of `quickFix` is something like only show quick fixes? If so, perhaps call it `onlyShowFixes`. This definition may then look something like this:

```
void showCompletions({bool autoInvoked: false, bool onlyShowFixes: false});
```

File: lib/completion.dart [hunk 11]
```diff
@@ -39,81 +39,102 @@ class DartCompleter extends CodeCompleter {
...
-        return new AnalysisCompletion(replaceOffset, replaceLength, completion);
-      }).toList();
-
-      List<Completion> completions =  analysisCompletions.map((completion) {
-        // TODO: Move to using a LabelProvider; decouple the data and rendering.
-        String displayString = completion.isMethod
-            ? '${completion.text}${completion.parameters}' : completion.text;
-        if (completion.isMethod && completion.returnType != null) {
-          displayString += ' → ${completion.returnType}';
+    if (editor.lookingForQuickFix) {
```
Comment (devoncarew) [inline lib/completion.dart]: Instead of storing `lookingForQuickFix` as a public field on the editor class, let's have it be passed into the CodeCompleter.complete() method as an optional named param.

File: lib/editing/editor.dart [hunk 12]
```diff
@@ -45,6 +48,8 @@ abstract class Editor {
 
   bool completionAutoInvoked;
 
+  bool lookingForQuickFix;
```
Comment (devoncarew) [inline lib/editing/editor.dart]: I think we can remove this field from the public API (I'll add some other comments about this).

File: lib/editing/editor.dart [hunk 13]
```diff
@@ -166,7 +183,9 @@ class Completion {
   /// completors. See [EditorFactory.supportsCompletionPositioning].
   final int cursorOffset;
 
-  Completion(this.value, {this.displayString, this.type, this.cursorOffset});
+  List<Edit> quickFixes = [];
```
Comment (devoncarew) [inline lib/editing/editor.dart]: Does an annotation need to know it's fixes?

File: lib/editing/editor_codemirror.dart [hunk 14]
```diff
@@ -166,6 +181,22 @@ class _CodeMirrorEditor extends Editor {
...
   }
 
+  void autoComplete({bool autoInvoked, bool quickFix}) {
+    if (autoInvoked != null && autoInvoked) {
+      completionAutoInvoked = true;
+    } else {
+      completionAutoInvoked = false;
+    }
+    if (quickFix != null && quickFix) {
+      lookingForQuickFix = true;
```
Comment (devoncarew) [inline lib/editing/editor_codemirror.dart]: Let's store this info instead in a private field - `_lookingForQuickFix`.

File: lib/editing/editor_codemirror.dart [hunk 15]
```diff
@@ -179,6 +210,12 @@ class _CodeMirrorEditor extends Editor {
   set completionAutoInvoked(bool value)
       => cm.jsProxy['state']['completionAutoInvoked'] = value;
 
+  bool get lookingForQuickFix
```
Comment (devoncarew) [inline lib/editing/editor_codemirror.dart]: I think this setter and getter pair can move to being just a regular private field - `_lookingForQuickFix`.

File: lib/playground.dart [hunk 16]
```diff
@@ -203,9 +203,14 @@ class Playground {
       _handleHelp();
     });
 
+    keys.bind(['alt-enter', 'ctrl-1'], (){
+      if (editor.document.hasIssueAtOffset) {
```
Comment (devoncarew) [inline lib/playground.dart]: If they hit ctrl-1, should we always look for quick-fixes? I don't think we need the conditional here.

File: lib/playground.dart [hunk 17]
```diff
@@ -579,6 +582,22 @@ ${info['libraryName'] == null ? "" : "**Library:** ${apiLink == null ? info['lib
         messageSpan.classes.add('message');
         messageSpan.text = issue.message;
         e.children.add(messageSpan);
+        source.offset = issue.charStart;
+        dartServices.fix(source).timeout(serviceCallTimeout).then((FixesResponse fixResponse) {
+          if (fixResponse.fixes.isNotEmpty) {
```
Comment (devoncarew) [inline lib/playground.dart]: I don't think I understand the flow here; my understanding is that:

- the user hits ctrl-1
- we send out a request for fixes
- we display them as code completions
- if the user selects one, we perform the indicated fix

But I think you may have a different flow in mind?

**Comment by devoncarew** [comment]
Lots of comments! And some confusion on my part. I think this PR will be simplified a bit by merging to master; it'll pick up some overlapping changes as well as the new services code.

File: lib/playground.dart (hunk 17 above)
Comment (kasperpeulen) [inline lib/playground.dart]: @devoncarew Lukechurch asked if I could add this. What this does is making sure the wrench is shown only if there is a quickfix available:

image

File: lib/editing/editor_codemirror.dart (hunk 15 above)
Comment (kasperpeulen) [inline lib/editing/editor_codemirror.dart]: I've tried storing information in the codemirror editor as fields, or private fiels, but this informatotion get lost, as there is new instance created of the editor at the beginning of the completionhelper. Therefore I think we must store it to the jsobject instead.

File: lib/dartservices_client/v1.dart (hunk 7 above)
Comment (kasperpeulen) [inline lib/dartservices_client/v1.dart]: Ok, that's fine, but if I put the code out now, the PR is not testable anymore, shall I leave it in for now, so you can still test it ? Then I will rebase the PR, once it is landed at dart-services and here in the master.

File: lib/services/common.dart (hunk 3 above)
Comment (kasperpeulen) [inline lib/services/common.dart]: Maybe leave it in, so that you guys test the PR, and then I rebase the PR when this is landed at dart-services ?

**Comment by kasperpeulen** [comment]
see #273

File: lib/playground.dart (hunk 17 above)
Comment (devoncarew) [inline lib/playground.dart]: Ah, ok. I understand now.

I don't believe we should be making service calls while we're building up the UI for analysis issues. Let's comment out or remove this code for now, and figure out if there's a solution that'll provide @lukechurch with the UI he wants.

It would be nice if the service call that sends us analysis issues, could provide a boolean on each issue telling us whether it had any quick fixes. This might be O(1) for the services layer to provide. I think a quick fix is always associated with a type of error.

File: lib/dartservices_client/v1.dart (hunk 7 above)
Comment (devoncarew) [inline lib/dartservices_client/v1.dart]: Rebasing once the new services file lands sounds good.

**Comment by devoncarew** [comment]
> I've tried storing information in the codemirror editor as fields, or private fiels, but this informatotion get lost, as there is new instance created of the editor at the beginning of the completionhelper. Therefore I think we must store it to the jsobject instead.

That sounds like a bug in the dart codemirror library, or our codemirror abstraction in dartpad. Possibly here: https://github.com/dart-lang/dart-pad/blob/master/lib/editing/editor_codemirror.dart#L98 ? I'll take a look -
